Update check/Python 3: use a new kernel32 wrapper for MoveFileEx instead of os.renames for renaming update installer when 'Postpone' button is clicked

josephsl opened the pull request:
### Link to issue number:
Fixes #9825 

### Summary of the issue:
In Python 3, if os.renames is attempted when moving files from one drive to another, error 17 (canot move file to a different drive) is presented.

### Description of how this pull request fixes the issue:
Changed to use shutil.move to "rename" the update installer when "Postpone" button is clicked. Note that because shutil.move blocks, a dedicated thread will be used to rename the installer.

### Testing performed:
Tested with a modified Python 3 staging source, checking to make sure the installer was renamed and update check state dictionary was consistent.

### Known issues with pull request:
None

### Change log entry:
None

leonardder commented:
I recall you said something about decreased performance.
Therefore, it might be beneficial to test the move action as follows

`winKernel.kernel32.MoveFileExW(source, dest, 0xa)`

josephsl commented:
Hi, it works, although because EndModal is called after this operation, there will be a delay between when Postpone button is clicked and dialog is gone. I’m hoping we can work on a solution, perhaps at a future time (say, 2020 and later). Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Thursday, June 27, 2019 11:08 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Author <author@noreply.github.com>
Subject: Re: [nvaccess/nvda] Update check/Python 3: use shutil.move instead of os.renames for renaming update installer when 'Postpone' button is clicked (#9836)

 

I recall you said something about decreased performance.
Therefore, it might be beneficial to test the move action as follows

winKernel.kernel32.MoveFileExW(source, dest, 0xa)

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/9836?email_source=notifications&email_token=AB4AXEAPKV3PMXI6BYLQVHDP4WTFHA5CNFSM4H4BQCRKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYZFCTI#issuecomment-506614093> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AB4AXECVDTQ6FXVYAP2DUPLP4WTFHANCNFSM4H4BQCRA> .

leonardder commented:
AH, setting the flag to 0x2 won't wait until the move action has 
completed, so then, the delay should disappear.

josephsl committed:
Update check/Python 3: use MoveFileEx instead of os.renames when copying the update installer after postponing an update. Re #9825.

A tip from Leonard de Ruijter (Babbage): because shutil.move introduces performance issues (namely when copying the update installer after clicking 'Postpone' button), use kernel32::MoveFileEx function with write-through (0x8) and copy allowed (0x2) flags set. This results in a slight delay after clicking Postpone button, but this gets the job done.

josephsl committed:
Update check/Python 3: use only copy allowed flag (0x2) when 'renaming' update installer. Re #9825.

Verified by Leonard de Ruijter (Babbage): setting just copy allowed flag (0x2) in kernel32::MoveFileEx restores performance.

LeonarddeR reviewed (changes requested):
I think I prefer adding a MoveFileEx wrapper function to winKernel that also raises WinError if appropriate. Also, please add a named constant for 0x2 as well as for the others, see the MoveFileEx docs.

josephsl committed:
WinKernel: update copyright headers and added module docstring. Re #9825.

josephsl committed:
WinKernel: add a wrapper and constants for kernel32::MoveFileEx. re #9825.

Reviewed by Leonard de Ruijter (Babbage): add a wrapper function for kernel32::MoveFileEx and add flag constants for this function. This will then allow any file move operations that relies on MoveFileEx to use this wrapper if appropriate.

josephsl committed:
Update check/Python 3: use just defined winKernel.moveFileEx when 'renaming' update installer after clicking 'Postpone' button. Re #9825.

leonardder commented on source/winKernel.py:
```diff
@@ -383,3 +385,16 @@ def forget(self):
+MOVEFILE_WRITE_THROUGH = 0x8
+
+def moveFileEx(lpExistingFileName, lpNewFileName, dwFlags):
+	res=kernel32.MoveFileExW(lpExistingFileName, lpNewFileName, dwFlags)
```
Res is not relevant to return here.

```suggestion
	if not kernel32.MoveFileExW(lpExistingFileName, lpNewFileName, dwFlags):
```

leonardder commented:
Could you look at installer.tryRemoveFile and the fileUtils module to use the new winKernel wrapper there?

josephsl commented:
Hi, actually, the return value is required for raising errors and to return how the function went in compliance with Windows API docs. Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Saturday, June 29, 2019 2:15 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Author <author@noreply.github.com>
Subject: Re: [nvaccess/nvda] Update check/Python 3: use a new kernel32 wrapper for MoveFileEx instead of os.renames for renaming update installer when 'Postpone' button is clicked (#9836)

 

@leonardder commented on this pull request.

  _____  

In source/winKernel.py <https://github.com/nvaccess/nvda/pull/9836#discussion_r298793694> :

> @@ -383,3 +385,16 @@ def forget(self):
 
               Necessary if you pass this HGLOBAL to an API that takes ownership and therefore will handle freeing itself.
 
               """
 
               self.value=None
 
+
 
+MOVEFILE_COPY_ALLOWED = 0x2
 
+MOVEFILE_CREATE_HARDLINK = 0x10
 
+MOVEFILE_DELAY_UNTIL_REBOOT = 0x4
 
+MOVEFILE_FAIL_IF_NOT_TRACKABLE = 0x20
 
+MOVEFILE_REPLACE_EXISTING = 0x1
 
+MOVEFILE_WRITE_THROUGH = 0x8
 
+
 
+def moveFileEx(lpExistingFileName, lpNewFileName, dwFlags):
 
+       res=kernel32.MoveFileExW(lpExistingFileName, lpNewFileName, dwFlags)
 

Res is not relevant to return here.

⬇️ Suggested change 

-       res=kernel32.MoveFileExW(lpExistingFileName, lpNewFileName, dwFlags)
 
+       if not kernel32.MoveFileExW(lpExistingFileName, lpNewFileName, dwFlags):
 

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/9836?email_source=notifications&email_token=AB4AXEDNTT6R36EOJOF6PMLP44RXTA5CNFSM4H4BQCRKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB5BLPXA#pullrequestreview-256030684> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AB4AXEDMSPNUKZ7QJDLURJLP44RXTANCNFSM4H4BQCRA> .

josephsl commented:
Hi, I can do this as part of this PR, or we can separate this into another PR so update check routine can serve as a canary before we modify more files. Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Saturday, June 29, 2019 2:22 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Author <author@noreply.github.com>
Subject: Re: [nvaccess/nvda] Update check/Python 3: use a new kernel32 wrapper for MoveFileEx instead of os.renames for renaming update installer when 'Postpone' button is clicked (#9836)

 

Could you look at installer.tryRemoveFile and the fileUtils module to use the new winKernel wrapper there?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/9836?email_source=notifications&email_token=AB4AXEEQX5OPDZ57DPPNQ3LP44SS5A5CNFSM4H4BQCRKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODY3VK4A#issuecomment-506942832> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AB4AXEAIVN6TXUFUARD56T3P44SS5ANCNFSM4H4BQCRA> .

josephsl commented:
Hi,

After thinking about Leonard's proposal, I think it would be best to postpone MoveFileEx wrapper writing until we gather more info on which modules/functions/classes can benefit from such a wrapper so it can be expanded beyond update check postpone button handling. Thus, I'm closing it, although the branch used will become a base for other MoveFileEx wrapper work.

Thanks.
